### PR TITLE
Fix issue with curlies being innapropriately unescaped in raw string literals.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/59603

Relates to test plan https://github.com/dotnet/roslyn/issues/55306

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 1]
```diff
@@ -153,13 +154,19 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
...
                                 var text = ((InterpolatedStringTextSyntax)content).TextToken.ValueText;
-                                builder.Add(new BoundLiteral(content, ConstantValue.Create(text, SpecialType.System_String), stringType));
+                                // Raw string literals have no escapes.  So there is no need to manipulate their value texts.
+                                if (!isRawInterpolatedString)
+                                {
+                                    text = ConstantValueUtils.UnescapeInterpolatedStringLiteral(text);
+                                }
+
+                                var constantValue = ConstantValue.Create(text, SpecialType.System_String);
+                                builder.Add(new BoundLiteral(content, constantValue, stringType));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:164]: the core cahnge here is that we always make the ConstantValue have the right actual text content.  THe unescaping only happens here when we make the boundliteral (and not for raw strings).  From this point on, nothing should touch this text.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 2]
```diff
@@ -748,7 +755,7 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
                     {
                         var boundLiteral = (BoundLiteral)part;
                         Debug.Assert(boundLiteral.ConstantValue != null && boundLiteral.ConstantValue.IsString);
-                        var literalText = ConstantValueUtils.UnescapeInterpolatedStringLiteral(boundLiteral.ConstantValue.StringValue);
+                        var literalText = boundLiteral.ConstantValue.StringValue;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:758]: intentionally removed.  We ensured at creation of hte bound nodes that hte literal constant values are correct.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs [hunk 3]
```diff
@@ -36,7 +37,7 @@ private void RenderAndVerify(string markup, string expectedOutput)
     private void RenderAndVerify(string markup, string expectedOutput, string? normalize)
     {
         var text = Render(markup, normalize);
-        ParseAllPrefixes(text);
+        // ParseAllPrefixes(text);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs:40]:
```suggestion
        ParseAllPrefixes(text);
```

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs [hunk 4]
```diff
@@ -4749,7 +4749,7 @@ public void InterpolationEscapeConstantValue_WithoutDefaultHandler()
           null
       IInterpolatedStringTextOperation (OperationKind.InterpolatedStringText, Type: null) (Syntax: ' }}')
         Text:
-          ILiteralOperation (OperationKind.Literal, Type: System.String, Constant: "" }}"", IsImplicit) (Syntax: ' }}')
+          ILiteralOperation (OperationKind.Literal, Type: System.String, Constant: "" }"", IsImplicit) (Syntax: ' }}')
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4752]: this test appears to be demonstrating that the bound nodes were in error previously.  For `$""{{ {i} }}""` the literals should only have one brace, not two like the test seemed to indicate.

**Comment by CyrusNajmabadi** [comment]
@jcouv for early thoughts.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs [hunk 5]
```diff
@@ -11386,6 +11386,44 @@ .maxstack  4
...
+    // Code size       43 (0x2b)
+    .maxstack  3
+    .locals init (CustomHandler V_0, //c
+                CustomHandler V_1)
+    IL_0000:  ldloca.s   V_1
+    IL_0002:  ldc.i4.s   9
+    IL_0004:  ldc.i4.0
+    IL_0005:  call       ""CustomHandler..ctor(int, int)""
+    IL_000a:  ldloca.s   V_1
+    IL_000c:  ldstr      ""{{1 + 2}}""
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs:11414]: not unescaped, as expected.

**Comment by CyrusNajmabadi** [comment]
@jcouv  @333fred @RikkiGibson  can you please take a look.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 6]
```diff
@@ -153,13 +154,23 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
...
-                                builder.Add(new BoundLiteral(content, ConstantValue.Create(text, SpecialType.System_String), stringType));
+                                // Raw string literals have no escapes.  So there is no need to manipulate their value texts.
+                                // We have to unescape normal interpolated strings as the parser stores their text without
+                                // interpretting {{ and }} sequences (as '{' and '}') respectively.  Changing that at the syntax
+                                // level might potentially be a breaking change, so we do the conversion here when creating the
+                                // bound nodes.
+                                if (!isRawInterpolatedString)
+                                {
+                                    text = ConstantValueUtils.UnescapeInterpolatedStringLiteral(text);
+                                }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:165]: note: i'm also amenable to actually changing the .ValueText for a normal interpolated string.  I was actually *super* surprised to discover that `{{` in a text chunk is treated as `{{` in the `.ValueText`!   Surprised to the point that it feels like a bug that we ended up working around in later portions of the code, as nothing else seems to do this.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 7]
```diff
@@ -153,13 +154,23 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
...
+                                builder.Add(new BoundLiteral(content, constantValue, stringType));
                                 if (isResultConstant)
                                 {
-                                    var constantVal = ConstantValue.Create(ConstantValueUtils.UnescapeInterpolatedStringLiteral(text), SpecialType.System_String);
-                                    resultConstant = (resultConstant is null)
-                                        ? constantVal
-                                        : FoldStringConcatenation(BinaryOperatorKind.StringConcatenation, resultConstant, constantVal);
+                                    resultConstant = resultConstant is null
+                                        ? constantValue
+                                        : FoldStringConcatenation(BinaryOperatorKind.StringConcatenation, resultConstant, constantValue);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:173]: note: the UnescapeInterpolatedStringLiteral is only called in this single place now.  i could move it to be a helper here

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs [hunk 8]
```diff
@@ -220,8 +220,9 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
                 if (fillin == null)
                 {
                     Debug.Assert(part is BoundLiteral && part.ConstantValue != null);
-                    // this is one of the literal parts
-                    stringBuilder.Append(part.ConstantValue.StringValue);
+                    // this is one of the literal parts.  If it contains a { or } then we need to escape those so that
+                    // they're treated the same way in string.Format.
+                    stringBuilder.Append(ConstantValueUtils.EscapeInterpolatedStringLiteral(part.ConstantValue.StringValue));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:225]: note: the EscapeInterpolatedStringLiteralis only called here. i could move that helper here instead :)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 9]
```diff
@@ -153,13 +154,23 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
                         case SyntaxKind.InterpolatedStringText:
                             {
                                 var text = ((InterpolatedStringTextSyntax)content).TextToken.ValueText;
-                                builder.Add(new BoundLiteral(content, ConstantValue.Create(text, SpecialType.System_String), stringType));
+                                // Raw string literals have no escapes.  So there is no need to manipulate their value texts.
+                                // We have to unescape normal interpolated strings as the parser stores their text without
+                                // interpretting {{ and }} sequences (as '{' and '}') respectively.  Changing that at the syntax
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:159]: Typo:
```suggestion
                                // interpreting {{ and }} sequences (as '{' and '}') respectively.  Changing that at the syntax
```

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 10]
```diff
@@ -77,19 +77,34 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
         {
             var builder = PooledStringBuilder.GetInstance();
             var stringBuilder = builder.Builder;
-            int formatLength = s.Length;
-            for (int i = 0; i < formatLength; i++)
+            for (int i = 0, formatLength = s.Length; i < formatLength; i++)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:80]: nit: I'd prefer reverting this change (ie. keeping the declaration for `int formatLength` above)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs [hunk 11]
```diff
@@ -11386,6 +11386,44 @@ .maxstack  4
 ");
     }
 
+    [Theory, WorkItem(59603, "https://github.com/dotnet/roslyn/issues/59603")]
+    [InlineData(@"$$$""""""{{1 + 2}}""""""")]
+    public void BracesNotEscaped(string expression)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs:11391]: Consider also adding the two related scenarios that David reported:
1. four dollars and two curlies
2. four dollars and three curlies

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs [hunk 12]
```diff
@@ -11386,6 +11386,44 @@ .maxstack  4
...
+    IL_0016:  ldloc.1
+    IL_0017:  stloc.0
+    IL_0018:  ldloca.s   V_0
+    IL_001a:  constrained. ""CustomHandler""
+    IL_0020:  callvirt   ""string object.ToString()""
+    IL_0025:  call       ""void System.Console.WriteLine(string)""
+    IL_002a:  ret
+}
+");
+    }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs:11425]: Could you add verification for IOperation for the added scenarios? (in this file is fine)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:225]: In my cursory search, I could only find one test affected by this (`InterpolatedStringWithCurlyBracesFollowerAfterFormatSpecifierTest`). Do we have others?
Feels like we should cover this more. For example, what if we have a backslash in the literal part before the curly? What if we have double curlies in the string literal?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 11)

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:80]: i can move it out, though this definitely seems more idiomatic as the normal way to iterate indices.  this way the 'formatLength' value isn't in a higher scope than necessary.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs:11391]: sure.  will do.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests_Handler.cs:11425]: sure!

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 13]
```diff
@@ -176,6 +187,24 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
...
+                    stringBuilder.Append(c);
+                    if (c is '{' or '}' &&
+                        i + 1 < formatLength &&
+                        value[i + 1] == c)
+                    {
+                        i++;
+                    }
+                }
+                return builder.ToStringAndFree();
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:207]: moved this as it's the only place we use this.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 14]
```diff
@@ -1242,6 +1240,314 @@ .maxstack  3
...
+#if NETCOREAPP
+        [InlineData(TargetFramework.Net60)]
+        [InlineData(TargetFramework.Net50)]
+#endif
+        [InlineData(TargetFramework.NetFramework)]
+        [InlineData(TargetFramework.NetStandard20)]
+        [InlineData(TargetFramework.Mscorlib461)]
+        [InlineData(TargetFramework.Mscorlib40)]
+        [Theory]
+        public void InterpolatedStringWithCurlyBracesAndAllStringValues(TargetFramework framework)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:1357]: added explicit tests to check the string.concat case as well as the string.format case.

**Review by 333fred — APPROVED** [review]
LGTM (commit 14)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs [hunk 15]
```diff
@@ -4749,12 +4749,89 @@ public void InterpolationEscapeConstantValue_WithoutDefaultHandler()
...
+  Parts(1):
+      IInterpolationOperation (OperationKind.Interpolation, Type: null) (Syntax: '{{i}}')
+        Expression:
+          ILocalReferenceOperation: i (OperationKind.LocalReference, Type: System.Int32) (Syntax: 'i')
+        Alignment:
+          null
+        FormatString:
+          null";
+
+            VerifyOperationTreeAndDiagnosticsForTest<InterpolatedStringExpressionSyntax>(new[] { code }, expectedOperationTree, expectedDiagnostics, parseOptions: TestOptions.RegularPreview);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4778]: nit: `parseOptions: TestOptions.RegularPreview` should no longer be needed (we've switched `CreateCompilation` to use that by default)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 14). Only minor comments left

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 16]
```diff
@@ -176,6 +187,24 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
 
             Debug.Assert(isResultConstant == (resultConstant != null));
             return new BoundUnconvertedInterpolatedString(node, builder.ToImmutableAndFree(), resultConstant, stringType);
+
+            static string unescapeInterpolatedStringLiteral(string value)
+            {
+                var builder = PooledStringBuilder.GetInstance();
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:193]: k

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:193]: > nit: could pass the Length of value as starting capacity.

note: PooledStringBuilder does not seem to support this concept.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs [hunk 17]
```diff
@@ -216,12 +216,12 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
             for (int i = 0; i <= n; i++)
             {
                 var part = node.Parts[i];
-                var fillin = part as BoundStringInsert;
-                if (fillin == null)
+                if (part is not BoundStringInsert fillin)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:219]: done.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4778]: done.

Comment (jcouv) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4778]: Did you push the commit?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 16)

**Pull request merged**